File: python/src/wordle_service.py
import asyncio
import uuid
from typing import Optional

from .types import (
    GameState,
    GameOptions,
    GuessResult,
    LetterCode,
    GameNotFoundError,
    GameOverError,
)
from .dictionary_service import DictionaryService


class WordleService:
    """
    Wordle Game Service

    Manages game state and validates guesses for Wordle games.
    """

    def __init__(self, dictionary: Optional[DictionaryService] = None):
        self._games: dict[str, GameState] = {}
        self._dictionary = dictionary or DictionaryService()

    def start_game(self, options: Optional[GameOptions] = None) -> str:
        """Start a new game and return the game ID."""
        if options is None:
            options = GameOptions()

        game_id = str(uuid.uuid4())
        answer = options.answer.upper() if options.answer else self._dictionary.get_random_word()
        max_guesses = options.max_guesses

        game = GameState(
            id=game_id,
            answer=answer,
            max_guesses=max_guesses,
        )

        self._games[game_id] = game
        return game_id

    async def submit_guess(self, game_id: str, guess: str) -> GuessResult:
        """Submit a guess for a game."""
        game = self._games.get(game_id)

        if game is None:
            raise GameNotFoundError(game_id)

        if game.won or game.lost:
            raise GameOverError(game_id)

        normalized_guess = guess.upper()

        await self._simulate_async_work()

        codes = self._calculate_letter_codes(normalized_guess, game.answer)

        game.guesses.append(normalized_guess)

        won = normalized_guess == game.answer
        lost = not won and len(game.guesses) >= game.max_guesses

        game.won = won
        game.lost = lost

        return GuessResult(
            guess=normalized_guess,
            codes=codes,
            remaining_guesses=game.max_guesses - len(game.guesses),
            won=won,
            lost=lost,
        )

    def get_game(self, game_id: str) -> Optional[GameState]:
        """Get current game state (for debugging)."""
        return self._games.get(game_id)

    def _calculate_letter_codes(self, guess: str, answer: str) -> list[LetterCode]:
        """Calculate letter codes for a guess."""
        codes = []
        for i, guess_char in enumerate(guess):
            if guess[i] == answer[i]:
                codes.append(LetterCode.GREEN)
            elif guess_char in answer:
                codes.append(LetterCode.YELLOW)
            else:
                codes.append(LetterCode.GREY)
        return codes

    async def _simulate_async_work(self) -> None:
        """Simulate async work (e.g., external validation, logging)."""
        await asyncio.sleep(0.01)

File: java/src/main/java/wordle/WordleService.java
package wordle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Wordle Game Service
 *
 * Manages game state and validates guesses for Wordle games.
 */
public class WordleService {

    private final Map<String, GameState> games = new HashMap<>();
    private final DictionaryService dictionary;

    public WordleService() {
        this.dictionary = new DictionaryService();
    }

    public WordleService(DictionaryService dictionary) {
        this.dictionary = dictionary;
    }

    /**
     * Start a new game.
     */
    public String startGame(GameOptions options) {
        String id = UUID.randomUUID().toString();
        String answer = (options.answer() != null)
            ? options.answer().toUpperCase()
            : dictionary.getRandomWord();
        int maxGuesses = options.maxGuesses() > 0 ? options.maxGuesses() : 6;

        games.put(id, new GameState(id, answer, maxGuesses));
        return id;
    }

    public String startGame() {
        return startGame(GameOptions.defaults());
    }

    /**
     * Submit a guess for a game.
     */
    public GuessResult submitGuess(String gameId, String guess) {
        GameState game = games.get(gameId);

        if (game == null) {
            throw new GameNotFoundException(gameId);
        }

        if (game.isWon() || game.isLost()) {
            throw new GameOverException(gameId);
        }

        String normalizedGuess = guess.toUpperCase();

        // Simulate async work (e.g., logging, external call)
        simulateAsyncWork();

        List<LetterCode> codes = calculateLetterCodes(normalizedGuess, game.getAnswer());
        game.addGuess(normalizedGuess);

        boolean won = normalizedGuess.equals(game.getAnswer());
        boolean lost = !won && game.getGuessCount() >= game.getMaxGuesses();

        game.setWon(won);
        game.setLost(lost);

        return new GuessResult(
            normalizedGuess,
            codes,
            game.getMaxGuesses() - game.getGuessCount(),
            won,
            lost
        );
    }

    /**
     * Get current game state (for debugging/inspection).
     */
    public GameState getGame(String gameId) {
        return games.get(gameId);
    }

    /**
     * Calculate letter codes for a guess against the answer.
     */
    private List<LetterCode> calculateLetterCodes(String guess, String answer) {
        List<LetterCode> codes = new ArrayList<>();

        for (int i = 0; i < guess.length(); i++) {
            char guessChar = guess.charAt(i);
            char answerChar = answer.charAt(i);

            if (guessChar == answerChar) {
                codes.add(LetterCode.GREEN);
            } else if (answer.indexOf(guessChar) >= 0) {
                codes.add(LetterCode.YELLOW);
            } else {
                codes.add(LetterCode.GREY);
            }
        }

        return codes;
    }

    private void simulateAsyncWork() {
        try {
            Thread.sleep(10);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
